Codegen all functions in a crate with global ASM as unimplemented

zhassan-aws opened the pull request:
### Description of changes: 

Currently, for crates that contain [module-level assembly](https://doc.rust-lang.org/beta/core/arch/macro.global_asm.html), codegen emits a warning, and skips the codegen of all `GlobalAsm` items. This may impact soundness as the assembly that we don't model could be affecting the state in that crate. This PR mitigates the soundness issue by codegening all functions in a crate as unimplemented if it has module-level assembly. This would cause verification to fail if _any_ function in that crate is called, with the following message:
```
Calling function "foo" defined in a crate with global_asm is not currently supported by Kani
```
and it would cause all other "SUCCESS" results to be reported as "UNDETERMINED".

The PR also leaves static variables in crates with global ASM undefined (only declared), which would cause a non-det value to be returned for them.

### Resolved issues:

Mitigates #316 


### Call-outs:

<!-- 
Address any potentially confusing code. Is there code added that needs to be cleaned up later? Is there code that is missing because it’s still in development? 
-->

### Testing:

* How is this change tested? One test added

* Is this a refactor change? No

### Checklist
- [X] Each commit message has a non-empty body, explaining why the change was made
- [X] Methods or procedures are documented
- [X] Regression or unit tests are included, or existing tests cover the modified code
- [X] My PR is restricted to a single feature or bugfix

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 and MIT licenses.

celinval reviewed (commented):
Can we implement all the functions using codegen_unimplemented as the only statement with a more friendly message?

zhassan-aws commented:
> Can we implement all the functions using codegen_unimplemented as the only statement with a more friendly message?

Done.

celinval commented on kani-compiler/src/codegen_cprover_gotoc/codegen/function.rs:
```diff
@@ -74,14 +74,18 @@ impl<'tcx> GotocCtx<'tcx> {
+        } else if self.should_skip_current_fn() || crate_has_global_asm {
             debug!("Skipping function {}", self.current_fn().readable_name());
+            let unsupported_operation = if crate_has_global_asm {
+                "Calling function \"".to_owned()
```
Also, you can just use `'` instead of `\"`. :)

zhassan-aws replied:
Good ideas :) Both done.

zhassan-aws commented:
> Can you add a quick check to your test please?
> 
> I was wondering what will happen if you define a generic function inside `crate_with_global_asm` but only use it inside the `global` crate.

Hmmm..  This seems problematic. I added the following function to `crate_with_global_asm`:
```rust
pub fn a_generic_function<T>(x: T) -> T {
    x
}
```
and added a harness that calls it from the main crate:
```rust
#[kani::proof]
fn call_generic_function_in_crate_with_global_asm() {
    let x = "Hello World";
    let y = crate_with_global_asm::a_generic_function(x);
    assert_eq!(y, "Hello World");
}
```
and verification succeeded :(

Is this what you were concerned about?

zhassan-aws commented:
Closing in favor of #1112
